File: felles/src/test/kotlin/no/nav/familie/kontrakter/felles/RessursTest.java
package no.nav.familie.kontrakter.felles;

import org.junit.Test;
import org.junit.jupiter.api.Assertions;

import java.time.LocalDate;

import static org.junit.Assert.assertEquals;

public class RessursTest {
    private LocalDate date = LocalDate.of(2019, 11, 30);

    @Test
    public void skal_lagre_object_på_ressurs_og_hente_opp_igjen() {

        Ressurs<TestObject> ressurs = Ressurs.Companion.success(new TestObject("tekst", 42L, date), "OK");

        TestObject result = ressurs.getData();
        assertEquals(result.nummer, 42);
        assertEquals(result.tekst, "tekst");
        assertEquals(result.date, date);
    }

    @Test
    public void skal_konvertere_success_til_json_string() {
        Ressurs ressurs = Ressurs.Companion.success(new TestObject("tekst", 42L, date), "OK");
        assertEquals(ressurs.toJson(), "{\"data\":{\"tekst\":\"tekst\",\"nummer\":42,\"date\":\"2019-11-30\"},\"status\":\"SUKSESS\",\"melding\":\"OK\",\"stacktrace\":null}");
    }

    @Test
    public void skal_konvertere_stacktrace_to_data() {
        Ressurs ressurs = Ressurs.Companion.failure("Dette er errormeldingen", new RuntimeException("Dette er feilkoden"));

        Assertions.assertTrue(ressurs.getStacktrace().contains("java.lang.RuntimeException: Dette er feilkoden\n" +
                "\tat no.nav.familie.kontrakter.felles.RessursTest.skal_konvertere_stacktrace_to_data"));
    }

    static class TestObject {
        private String tekst;
        private long nummer;
        private LocalDate date;

        public TestObject() {
        }

        public TestObject(String tekst, long nummer, LocalDate date) {
            this.tekst = tekst;
            this.nummer = nummer;
            this.date = date;
        }
    }
}

File: felles/src/test/kotlin/no/nav/familie/kontrakter/felles/RessursTest.kt
package no.nav.familie.kontrakter.felles

import no.nav.familie.kontrakter.felles.Ressurs.Companion.failure
import no.nav.familie.kontrakter.felles.Ressurs.Companion.success
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import tools.jackson.databind.json.JsonMapper
import java.time.LocalDate

class RessursTest {
    private val date = LocalDate.of(2019, 11, 30)

    @Test
    fun `skal lagre object på ressurs og hente opp igjen`() {
        val (result) = success(TestObject("tekst", 42L, date), "OK")
        Assertions.assertEquals(result?.nummer, 42L)
        Assertions.assertEquals(result?.tekst, "tekst")
        Assertions.assertEquals(result?.date, date)
    }

    @Test
    fun `skal konvertere success til json string`() {
        val ressurs: Ressurs<*> = success(TestObject("tekst", 42L, date), "OK")
        Assertions.assertEquals(
            ressurs.toJson(),
            "{\"data\":{\"tekst\":\"tekst\",\"nummer\":42,\"date\":\"2019-11-30\"}," +
                "\"status\":\"SUKSESS\",\"melding\":\"OK\",\"frontendFeilmelding\":null,\"stacktrace\":null,\"callId\":null}",
        )
    }

    @Test
    fun `skal konvertere success til json string med vanlig objektmapper`() {
        val ressurs: Ressurs<*> = success("a", "OK")
        Assertions.assertEquals(
            "{\"data\":\"a\",\"status\":\"SUKSESS\",\"melding\":\"OK\",\"frontendFeilmelding\":null,\"stacktrace\":null,\"callId\":null}",
            JsonMapper()
                .writeValueAsString(ressurs),
        )
    }

    @Test
    fun `skal konvertere stacktrace to data`() {
        val (_, _, _, _, stacktrace) =
            failure<Any>(errorMessage = "Dette er errormeldingen", error = RuntimeException("Dette er feilkoden"))
        Assertions.assertTrue(
            stacktrace!!.contains(
                "java.lang.RuntimeException: Dette er feilkoden" +
                    System.getProperty("line.separator") +
                    "\tat no.nav.familie.kontrakter.felles.RessursTest.skal konvertere stacktrace to data",
            ),
        )
    }

    internal class TestObject(
        val tekst: String,
        val nummer: Long,
        val date: LocalDate?,
    )
}
